[bug]: Errors encountered during `uv run pytest` on Windows

### Summary

Hello,

I encountered the following errors during `uv run pytest` and wanted to report them with the relevant traceback for reference.

I ran it in Command Prompt as an **administrator**.

## ===short test summary info===
```
FAILED packages/memtomem/tests/test_cli_index_noop_e2e.py::TestFreshNoopIndexSubprocess::test_init_index_search_via_subprocess - TypeError: argument of type 'NoneType' is not iterable
FAILED packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries - AssertionError: assert 11 == 20
FAILED packages/memtomem/tests/test_uninstall_cmd.py::TestRuntimePidCleanedWithOther::test_runtime_pid_deleted_and_subdir_pruned - AssertionError: empty runtime subdir must be pruned
FAILED packages/memtomem/tests/test_wiki_store.py::TestIsDirty::test_dirty_with_modified_tracked_file - UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 196: illegal multibyte sequence
```


## ===FAILURES===
```
____________________________________________________________________________________ TestFreshNoopIndexSubprocess.test_init_index_search_via_subprocess ____________________________________________________________________________________

self = <tests.test_cli_index_noop_e2e.TestFreshNoopIndexSubprocess object at 0x0000020DDF615F90>, tmp_path = WindowsPath('C:/Users/user/AppData/Local/Temp/pytest-of-user/pytest-46/test_init_index_search_via_sub0')

    def test_init_index_search_via_subprocess(self, tmp_path):
        """Out-of-process variant: catches regressions that only manifest
        across the HOME / XDG boundary (e.g. config.json path resolution).

        Uses the ``mm`` script installed by ``uv pip install -e`` (co-located
        with ``sys.executable``) rather than ``python -m memtomem`` — the
        package has no ``__main__`` module, and the installed entry point is
        what real users hit.
        """
        # ``shutil.which`` adds the platform-correct suffix (``.exe`` on
        # Windows via PATHEXT, none on POSIX), so the same lookup works
        # against both ``.venv/bin/mm`` and ``.venv/Scripts/mm.exe``.
        bin_dir = os.path.dirname(sys.executable)
        mm_bin = shutil.which("mm", path=bin_dir)
        # Fail loudly instead of pytest.skip — any valid test environment
        # (``uv run pytest`` or ``uv pip install -e``) must provide the
        # ``mm`` entry point. A silent skip here would turn this subprocess
        # regression guard into CI false-green if the editable install is
        # ever dropped.
        if mm_bin is None:
            pytest.fail(
                f"mm binary not found in {bin_dir}. "
                "Run `uv pip install -e packages/memtomem[all]` before testing."
            )

        home = tmp_path / "home"
        home.mkdir()
        mem_dir = _make_memory_dir(str(home))

        env = os.environ.copy()
        env["HOME"] = str(home)
        env["XDG_CONFIG_HOME"] = str(home / ".config")

        def _run(*args: str) -> subprocess.CompletedProcess:
            return subprocess.run(
                [mm_bin, *args],
                env=env,
                capture_output=True,
                text=True,
                timeout=60,
            )

        r = _run(
            "init",
            "-y",
            "--provider",
            "none",
            "--memory-dir",
            mem_dir,
            "--mcp",
            "skip",
        )
        assert r.returncode == 0, f"init failed:\nstdout={r.stdout}\nstderr={r.stderr}"

        r = _run("index", mem_dir)
        assert r.returncode == 0, f"index failed:\nstdout={r.stdout}\nstderr={r.stderr}"
        assert "no such table" not in (r.stdout + r.stderr)
        assert "1 new" in r.stdout or "1 file" in r.stdout

        r = _run("search", "hello")
        assert r.returncode == 0, f"search failed:\nstdout={r.stdout}\nstderr={r.stderr}"
>       assert "hello world" in r.stdout
               ^^^^^^^^^^^^^^^^^^^^^^^^^
E       TypeError: argument of type 'NoneType' is not iterable

packages\memtomem\tests\test_cli_index_noop_e2e.py:153: TypeError
_______________________________________________________________________________ TestPersistenceAndConcurrency.test_concurrent_enqueue_does_not_lose_entries ________________________________________________________________________________

self = <tests.test_index_debounce.TestPersistenceAndConcurrency object at 0x0000020DE13AE990>
queue_file = WindowsPath('C:/Users/user/AppData/Local/Temp/pytest-of-user/pytest-46/test_concurrent_enqueue_does_n0/index_debounce_queue.json')

    def test_concurrent_enqueue_does_not_lose_entries(self, queue_file: Path) -> None:
        """Two threads enqueue distinct paths in parallel. The flock guarantees
        both writes land. Without the lock, the second writer's load+save
        would clobber the first writer's entry."""
        threads: list[threading.Thread] = []
        for i in range(20):
            t = threading.Thread(
                target=debounce.enqueue,
                args=(f"/tmp/file_{i:02d}.py",),
                kwargs={"now": 100.0 + i, "queue_file": queue_file},
            )
            threads.append(t)
        for t in threads:
            t.start()
        for t in threads:
            t.join()
        entries = _read_raw(queue_file)["entries"]
>       assert len(entries) == 20
E       AssertionError: assert 11 == 20
E        +  where 11 = len({'/tmp/file_00.py': {'first_seen': 100.0, 'force': False, 'last_seen': 100.0, 'namespace': None}, '/tmp/file_01.py': {...amespace': None}, '/tmp/file_05.py': {'first_seen': 105.0, 'force': False, 'last_seen': 105.0, 'namespace': None}, ...})

packages\memtomem\tests\test_index_debounce.py:277: AssertionError
________________________________________________________________________________ TestRuntimePidCleanedWithOther.test_runtime_pid_deleted_and_subdir_pruned _________________________________________________________________________________

self = <tests.test_uninstall_cmd.TestRuntimePidCleanedWithOther object at 0x0000020DE3164B90>, home = WindowsPath('C:/Users/user/AppData/Local/Temp/pytest-of-user/pytest-46/test_runtime_pid_deleted_and_s0/home')

    def test_runtime_pid_deleted_and_subdir_pruned(self, home):
        from memtomem._runtime_paths import ensure_runtime_dir, runtime_dir

        _seed_state(home)
        rt = ensure_runtime_dir()
        pid_file = rt / "server.pid"
        pid_file.write_text("0", encoding="utf-8")

        result = CliRunner().invoke(cli, ["uninstall", "-y"])

        assert result.exit_code == 0, result.output
        assert not pid_file.exists(), "runtime pid file must be deleted"
        # subdir should be gone too — we own it
>       assert not runtime_dir().exists(), "empty runtime subdir must be pruned"
E       AssertionError: empty runtime subdir must be pruned
E       assert not True
E        +  where True = exists()
E        +    where exists = WindowsPath('C:/Users/TONYST~1/AppData/Local/Temp/memtomem-0').exists
E        +      where WindowsPath('C:/Users/TONYST~1/AppData/Local/Temp/memtomem-0') = <function runtime_dir at 0x0000020DDF4AF240>()

packages\memtomem\tests\test_uninstall_cmd.py:470: AssertionError
------------------------------------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------------------------------------
INFO     memtomem.config:config.py:1511 Migrated auto_discover -> explicit memory_dirs: added 0 provider path(s). See C:\Users\user\AppData\Local\Temp\pytest-of-user\pytest-46\test_runtime_pid_deleted_and_s0\home\.memtomem\config.json (auto_discover is deprecated and will be removed).
____________________________________________________________________________________________ TestIsDirty.test_dirty_with_modified_tracked_file _____________________________________________________________________________________________

self = <tests.test_wiki_store.TestIsDirty object at 0x0000020DE3B23BB0>, wiki_root = WindowsPath('C:/Users/user/AppData/Local/Temp/pytest-of-user/pytest-46/test_dirty_with_modified_track0/wiki')

    def test_dirty_with_modified_tracked_file(self, wiki_root: Path) -> None:
        """A modification to a tracked file flips ``is_dirty`` to True."""
        store = WikiStore.at_default()
        store.init()
        readme = wiki_root / "README.md"
>       readme.write_text(readme.read_text() + "\nlocal note\n", encoding="utf-8")
                          ^^^^^^^^^^^^^^^^^^

packages\memtomem\tests\test_wiki_store.py:214:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\pathlib\_local.py:546: in read_text
    return PathBase.read_text(self, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = WindowsPath('C:/Users/user/AppData/Local/Temp/pytest-of-user/pytest-46/test_dirty_with_modified_track0/wiki/README.md'), encoding = 'locale', errors = None, newline = None

    def read_text(self, encoding=None, errors=None, newline=None):
        """
        Open the file in text mode, read it, and close the file.
        """
        with self.open(mode='r', encoding=encoding, errors=errors, newline=newline) as f:
>           return f.read()
                   ^^^^^^^^
E           UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 196: illegal multibyte sequence

..\..\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\pathlib\_abc.py:633: UnicodeDecodeError
============================================================================================================= warnings summary =============================================================================================================
packages/memtomem/tests/test_cli_index_noop_e2e.py::TestFreshNoopIndexSubprocess::test_init_index_search_via_subprocess
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-8 (_readerthread)

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\subprocess.py", line 1615, in _readerthread
      buffer.append(fh.read())
                    ~~~~~~~^^
  UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 261: illegal multibyte sequence

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_cli_index_noop_e2e.py::TestFreshNoopIndexSubprocess::test_init_index_search_via_subprocess
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-4 (_readerthread)

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\subprocess.py", line 1615, in _readerthread
      buffer.append(fh.read())
                    ~~~~~~~^^
  UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 21: illegal multibyte sequence

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_export_import_roundtrip.py: 9 warnings
packages/memtomem/tests/test_golden_path.py: 2 warnings
  C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\embedding\onnx.py:90: UserWarning: The model sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 now uses mean pooling instead of CLS embedding. In order to preserve the previous behaviour, consider either pinning fastembed version to 0.5.1 or using `add_custom_model` functionality.
    self._model = TextEmbedding(

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1304 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1302 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1300 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1305 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1309 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1301 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1299 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1296 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_index_debounce.py::TestPersistenceAndConcurrency::test_concurrent_enqueue_does_not_lose_entries
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\_pytest\threadexception.py:58: PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1297 (enqueue)

  Traceback (most recent call last):
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 223, in lock
      msvcrt.locking(  # type: ignore[attr-defined]
      ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          fd,
          ^^^
          mode,
          ^^^^^
          self._msvcrt_lock_length,
          ^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  OSError: [Errno 36] Resource deadlock avoided

  The above exception was the direct cause of the following exception:

  Traceback (most recent call last):
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 1043, in _bootstrap_inner
      self.run()
      ~~~~~~~~^^
    File "C:\Users\user\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\threading.py", line 994, in run
      self._target(*self._args, **self._kwargs)
      ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 193, in enqueue
      with _Lock(qp):
           ~~~~~^^^^
    File "C:\Users\user\memtomem-test\memtomem\packages\memtomem\src\memtomem\indexing\debounce.py", line 169, in __enter__
      portalocker.lock(self._fp, portalocker.LOCK_EX)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 316, in lock
      locker(file, flags)
      ~~~~~~^^^^^^^^^^^^^
    File "C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\portalocker\portalocker.py", line 230, in lock
      raise exceptions.AlreadyLocked(
      ...<3 lines>...
      ) from exc_value
  portalocker.exceptions.AlreadyLocked: (1, '[Errno 36] Resource deadlock avoided')

  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
    warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))

packages/memtomem/tests/test_webhooks.py::TestConfigSections::test_rerank_config_validation
  C:\Users\user\memtomem-test\memtomem\.venv\Lib\site-packages\pydantic\main.py:250: DeprecationWarning: rerank.top_k is deprecated; migrating to rerank.min_pool. Use rerank.oversample + rerank.min_pool + rerank.max_pool to scale the reranker candidate pool with the request top_k. (Slated for removal in memtomem 0.3.)
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

### Surface

Other

### Steps to reproduce

`uv run pytest` on Windows

### Expected behavior

0 failed

### Actual behavior

4 failed

### memtomem version

0.1.35

### Install method

From source (editable / dev)

### Python version

3.13.5

### Operating system

Windows

### MCP client (if applicable)

None

### Additional context

_No response_

I’m reporting the additional related error messages.

In particular, the issue where `tokenizer_config.json` cannot be found seems quite unusual. I may be mistaken, as I have been trying various things and could be mixing things up, but it appears that the presence of the file is inconsistent — sometimes it seems to exist, and sometimes it does not.

I could modify the code myself and submit a PR, but if there are any related internal changes that also need to be made, I thought it would be better to discuss them first before proceeding. So I wanted to report this first.

## === short test summary info ===
```
FAILED packages/memtomem/tests/test_embedding_progress.py::test_onnx_numerical_parity_with_and_without_progress[asyncio] - memtomem.errors.EmbeddingError: ONNX embedding failed: Could not find tokenizer_config.json in C:\Users\user\....
FAILED packages/memtomem/tests/test_privacy_long_content.py::TestLongContentScan::test_one_megabyte_scan_under_perf_ceiling - AssertionError: 1MB scan took 223.5 ms — exceeds the 200 ms ceiling, suggesting a non-linear regression
```

## === FAILURES ===
```
____________________________ test_onnx_numerical_parity_with_and_without_progress[asyncio] ____________________________

self = <memtomem.embedding.onnx.OnnxEmbedder object at 0x000001B45AD39C70>
texts = ['short one', 'this is a slightly longer sentence to skew padding shape', 'tiny', 'another medium length sentence with some content to embed', 'x', 'a final reasonably long sentence ending the parity probe set']

    async def embed_texts(
        self,
        texts: Sequence[str],
        *,
        on_progress: Callable[[int, int], None] | None = None,
    ) -> list[list[float]]:
        if not texts:
            return []
        text_list = list(texts)
        total = len(text_list)

        if on_progress is None:
            # Fast path — no callback plumbing, no cross-thread hops.
            try:
>               return await asyncio.to_thread(self._embed_sync, text_list, None)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

packages\memtomem\src\memtomem\embedding\onnx.py:152:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\asyncio\threads.py:25: in to_thread
    return await loop.run_in_executor(None, func_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
..\..\AppData\Roaming\uv\python\cpython-3.13.5-windows-x86_64-none\Lib\concurrent\futures\thread.py:59: in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
packages\memtomem\src\memtomem\embedding\onnx.py:120: in _embed_sync
    model = self._get_model()
            ^^^^^^^^^^^^^^^^^
packages\memtomem\src\memtomem\embedding\onnx.py:90: in _get_model
    self._model = TextEmbedding(
.venv\Lib\site-packages\fastembed\text\text_embedding.py:114: in __init__
    self.model = EMBEDDING_MODEL_TYPE(
.venv\Lib\site-packages\fastembed\text\onnx_embedding.py:260: in __init__
    self.load_onnx_model()
.venv\Lib\site-packages\fastembed\text\onnx_embedding.py:325: in load_onnx_model
    self._load_onnx_model(
.venv\Lib\site-packages\fastembed\text\onnx_text_model.py:68: in _load_onnx_model
    self.tokenizer, self.special_token_to_id = load_tokenizer(model_dir=model_dir)
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

model_dir = WindowsPath('C:/Users/user/.memtomem/cache/fastembed/models--qdrant--all-MiniLM-L6-v2-onnx/snapshots/5f1b8cd78bc4fb444dd171e59b18f3a3af89a079')

    def load_tokenizer(model_dir: Path) -> tuple[Tokenizer, dict[str, int]]:
        config_path = model_dir / "config.json"
        if not config_path.exists():
            raise ValueError(f"Could not find config.json in {model_dir}")

        tokenizer_path = model_dir / "tokenizer.json"
        if not tokenizer_path.exists():
            raise ValueError(f"Could not find tokenizer.json in {model_dir}")

        tokenizer_config_path = model_dir / "tokenizer_config.json"
        if not tokenizer_config_path.exists():
>           raise ValueError(f"Could not find tokenizer_config.json in {model_dir}")
E           ValueError: Could not find tokenizer_config.json in C:\Users\user\.memtomem\cache\fastembed\models--qdrant--all-MiniLM-L6-v2-onnx\snapshots\5f1b8cd78bc4fb444dd171e59b18f3a3af89a079

.venv\Lib\site-packages\fastembed\common\preprocessor_utils.py:32: ValueError

The above exception was the direct cause of the following exception:

    @pytest.mark.anyio
    async def test_onnx_numerical_parity_with_and_without_progress():
        """The progress-streaming code path must produce identical vectors
        to the no-progress fast path on the same input. Catches accidental
        drift if a future refactor reorders the per-yield iteration or
        introduces an inadvertent transformation in the callback wrap.

        Skipped without fastembed; runs in the ``golden-path (ONNX bge-m3)``
        CI job and any local env with fastembed installed.
        """
        pytest.importorskip("fastembed")
        pytest.importorskip("numpy")
        import numpy as np

        config = _onnx_config(model="all-MiniLM-L6-v2", dimension=384)
        emb_a = OnnxEmbedder(config)
        emb_b = OnnxEmbedder(config)
        texts = [
            "short one",
            "this is a slightly longer sentence to skew padding shape",
            "tiny",
            "another medium length sentence with some content to embed",
            "x",
            "a final reasonably long sentence ending the parity probe set",
        ]
        try:
>           vecs_no_progress = await emb_a.embed_texts(texts)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

packages\memtomem\tests\test_embedding_progress.py:377:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <memtomem.embedding.onnx.OnnxEmbedder object at 0x000001B45AD39C70>
texts = ['short one', 'this is a slightly longer sentence to skew padding shape', 'tiny', 'another medium length sentence with some content to embed', 'x', 'a final reasonably long sentence ending the parity probe set']

    async def embed_texts(
        self,
        texts: Sequence[str],
        *,
        on_progress: Callable[[int, int], None] | None = None,
    ) -> list[list[float]]:
        if not texts:
            return []
        text_list = list(texts)
        total = len(text_list)

        if on_progress is None:
            # Fast path — no callback plumbing, no cross-thread hops.
            try:
                return await asyncio.to_thread(self._embed_sync, text_list, None)
            except EmbeddingError:
                raise
            except Exception as exc:
>               raise EmbeddingError(f"ONNX embedding failed: {exc}") from exc
E               memtomem.errors.EmbeddingError: ONNX embedding failed: Could not find tokenizer_config.json in C:\Users\user\.memtomem\cache\fastembed\models--qdrant--all-MiniLM-L6-v2-onnx\snapshots\5f1b8cd78bc4fb444dd171e59b18f3a3af89a079

packages\memtomem\src\memtomem\embedding\onnx.py:156: EmbeddingError
------------------------------------------------ Captured stderr call -------------------------------------------------
[05/06/26 22:56:15] INFO     Loading ONNX embedding model            onnx.py:81
                             sentence-transformers/all-MiniLM-L6-v2
                             (threads=4,
                             cache_dir=C:\Users\user\.memtomem\
                             cache\fastembed) …
-------------------------------------------------- Captured log call --------------------------------------------------
INFO     memtomem.embedding.onnx:onnx.py:81 Loading ONNX embedding model sentence-transformers/all-MiniLM-L6-v2 (threads=4, cache_dir=C:\Users\user\.memtomem\cache\fastembed) …

____________________________ TestLongContentScan.test_one_megabyte_scan_under_perf_ceiling ____________________________

self = <tests.test_privacy_long_content.TestLongContentScan object at 0x000001B459B18D50>

    def test_one_megabyte_scan_under_perf_ceiling(self):
        # Soft ceiling — not a microbenchmark, just a quadratic-regression
        # guard. The 9 current patterns are short prefix-anchored regexes
        # and ``re.finditer`` over 1 MB completes in single-digit ms on
        # CI hardware; 200 ms gives ample headroom for jitter while
        # still catching an accidental O(N^2) rewrite.
        text = ("a" * 999_980) + _SECRET
        start = time.perf_counter()
        hits = privacy.scan(text)
        elapsed = time.perf_counter() - start
        assert hits, "Sanity check: the secret must still be detected"
>       assert elapsed < 0.2, (
            f"1MB scan took {elapsed * 1000:.1f} ms — exceeds the 200 ms "
            "ceiling, suggesting a non-linear regression"
        )
E       AssertionError: 1MB scan took 223.5 ms — exceeds the 200 ms ceiling, suggesting a non-linear regression
E       assert 0.22345239999958721 < 0.2

packages\memtomem\tests\test_privacy_long_content.py:121: AssertionError
```

--- a/packages/memtomem/src/memtomem/indexing/debounce.py
+++ b/packages/memtomem/src/memtomem/indexing/debounce.py
@@ -38,6 +38,7 @@ import json
 import logging
 import os
 import tempfile
+import threading
 import time
 from dataclasses import asdict, dataclass, field
 from pathlib import Path
@@ -139,6 +140,33 @@ def _save(path: Path, entries: dict[str, QueueEntry]) -> None:
         raise
 
 
+# Per-lockfile ``threading.Lock`` for intra-process serialization. The
+# file lock (``portalocker.lock``) is the cross-process barrier, but on
+# Windows ``LockFileEx`` does not reliably block a *second handle* from
+# the *same* process holding the file open through ``open(..., "a+b")`` —
+# two threads in one Python process each opening the sidecar lockfile
+# can both pass ``portalocker.lock(LOCK_EX)`` and race the read-modify-
+# write of the queue, losing entries (#759 failure 2). Holding a
+# threading.Lock keyed to the lockfile path before acquiring the file
+# lock collapses same-process contention to a single waiter, so
+# portalocker only ever sees one handle per process competing.
+_intra_process_locks: dict[Path, threading.Lock] = {}
+_intra_process_locks_guard = threading.Lock()
+
+
+def _intra_process_lock_for(path: Path) -> threading.Lock:
+    """Return the threading.Lock associated with ``path``, creating it
+    on first use. The dict accumulates one entry per distinct lockfile
+    path observed in the process — bounded in practice by the number of
+    queue files (one in normal usage)."""
+    with _intra_process_locks_guard:
+        lock = _intra_process_locks.get(path)
+        if lock is None:
+            lock = threading.Lock()
+            _intra_process_locks[path] = lock
+        return lock
+
+
 class _Lock:
     """``portalocker.lock(LOCK_EX)`` on a sidecar lockfile next to the queue.
 
@@ -151,30 +179,51 @@ class _Lock:
 
     The sidecar (``.<queue_name>.lock``) is never replaced; every
     process locks the same inode for the duration of its critical
-    section, so serialization is correct.
+    section, so serialization is correct across processes.
 
-    Cross-platform via ``portalocker``; concurrent callers serialize
-    on Windows the same as POSIX (replaces the prior Windows-no-op
-    fallback that relied on hooks firing serially per Claude Code
-    session — see #625).
+    Two-layer locking (#759):
+
+    - **Intra-process** ``threading.Lock`` keyed by the lockfile path
+      (``_intra_process_lock_for``). Acquired first; serializes threads
+      inside a single Python process before any file-handle work.
+      Required because Windows ``LockFileEx`` does not reliably block a
+      second handle from the same process — without this layer, a
+      multi-threaded plugin host (e.g. Claude Code's bursty ``Write``
+      hook fanout) loses queue entries on Windows.
+    - **Cross-process** ``portalocker.lock(LOCK_EX)`` on the sidecar
+      lockfile. Acquired second; serializes parallel ``mm index
+      --debounce-window`` invocations that don't share Python state.
+
+    Both layers release in reverse order in ``__exit__``.
     """
 
     def __init__(self, path: Path) -> None:
         self._lock_path = path.parent / f".{path.name}.lock"
+        self._intra_lock = _intra_process_lock_for(self._lock_path)
         self._fp: IO[bytes] | None = None
 
     def __enter__(self) -> "_Lock":
-        self._lock_path.parent.mkdir(parents=True, exist_ok=True)
-        self._fp = open(self._lock_path, "a+b")
-        portalocker.lock(self._fp, portalocker.LOCK_EX)
+        self._intra_lock.acquire()
+        try:
+            self._lock_path.parent.mkdir(parents=True, exist_ok=True)
+            self._fp = open(self._lock_path, "a+b")
+            portalocker.lock(self._fp, portalocker.LOCK_EX)
+        except BaseException:
+            # File-lock acquisition failed — must release the intra
+            # lock so other threads aren't permanently blocked.
+            self._intra_lock.release()
+            raise
         return self
 
     def __exit__(self, *exc) -> None:
-        if self._fp is None:
-            return
-        portalocker.unlock(self._fp)
-        self._fp.close()
-        self._fp = None
+        try:
+            if self._fp is None:
+                return
+            portalocker.unlock(self._fp)
+            self._fp.close()
+            self._fp = None
+        finally:
+            self._intra_lock.release()
 
 
 def enqueue(

--- a/packages/memtomem/tests/test_index_debounce.py
+++ b/packages/memtomem/tests/test_index_debounce.py
@@ -276,6 +276,40 @@ class TestPersistenceAndConcurrency:
         entries = _read_raw(queue_file)["entries"]
         assert len(entries) == 20
 
+    def test_intra_process_lock_serializes_threads_when_file_lock_is_noop(
+        self, queue_file: Path, monkeypatch
+    ) -> None:
+        """Pin the in-process ``threading.Lock`` independently of the
+        cross-process file lock.
+
+        On Windows, ``LockFileEx`` is unreliable between two handles in
+        the *same* process — if the file lock is the only barrier,
+        same-process threads race the read-modify-write and lose
+        entries (#759 failure 2). By stubbing ``portalocker.lock`` /
+        ``unlock`` to no-ops, this test simulates that regression on
+        every platform: the test passes only because the intra-process
+        ``threading.Lock`` is the actual serializer. Removing the
+        threading.Lock layer from ``_Lock`` flips this to a flaky
+        ``len(entries) < 20`` failure.
+        """
+        monkeypatch.setattr(debounce.portalocker, "lock", lambda *a, **kw: None)
+        monkeypatch.setattr(debounce.portalocker, "unlock", lambda *a, **kw: None)
+
+        threads: list[threading.Thread] = []
+        for i in range(20):
+            t = threading.Thread(
+                target=debounce.enqueue,
+                args=(f"/tmp/file_{i:02d}.py",),
+                kwargs={"now": 100.0 + i, "queue_file": queue_file},
+            )
+            threads.append(t)
+        for t in threads:
+            t.start()
+        for t in threads:
+            t.join()
+        entries = _read_raw(queue_file)["entries"]
+        assert len(entries) == 20
+
     def test_partial_drain_writes_remaining_back(self, queue_file: Path) -> None:
         """Two entries, one indexer-errors. The successful one is gone from
         disk; the failing one remains so the next hook call retries."""

